exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		infraMachineNameToMachine[infraRef.Name] = machine
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		ownerRefs := infraMachine.GetOwnerReferences()
+
+		machine, ok := infraMachineNameToMachine[infraMachine.GetName()]
+		if !ok {
+			return errors.Errorf("failed to patch ownerRef for infraMachine %q because no Machine has an infraRef pointing to it", infraMachine.GetName())
+		}
+		machineRef := metav1.NewControllerRef(&machine, machine.GroupVersionKind())
+		if !util.HasOwnerRef(ownerRefs, *machineRef) {
+			log.V(2).Info("Setting ownerRef on infraMachine", "infraMachine", infraMachine.GetName(), "namespace", infraMachine.GetNamespace(), "machine", machine.GetName())
+			ownerRefs = util.EnsureOwnerRef(ownerRefs, *machineRef)
+			infraMachine.SetOwnerReferences(ownerRefs)
+			if err := r.Client.Update(ctx, infraMachine); err != nil {
+				return errors.Wrapf(err, "failed to update infraMachine %q in namespace %q", infraMachine.GetName(), infraMachine.GetNamespace())
+			}
+		}
+	}
+
+	return nil
+}
+
+// getNewMachine creates a new Machine object. The name of the newly created resource is going
+// to be created by the API server, we set the generateName field.
+func getNewMachine(mp *expv1.MachinePool, infraMachine *unstructured.Unstructured) *clusterv1.Machine {
+	infraRef := corev1.ObjectReference{}
+	if infraMachine != nil {
+		infraRef.APIVersion = infraMachine.GetAPIVersion()
+		infraRef.Kind = infraMachine.GetKind()
+		infraRef.Name = infraMachine.GetName()
+		infraRef.Namespace = infraMachine.GetNamespace()
+	}
+
+	annotations := mp.Spec.Template.Annotations
+
+	machine := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", mp.Name),
+			// Note: by setting the ownerRef on creation we signal to the Machine controller that this is not a stand-alone Machine.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(mp, mp.GroupVersionKind())},
+			Namespace:       mp.Namespace,
+			Labels:          make(map[string]string),
+			Annotations:     annotations,
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName:       mp.Spec.ClusterName,
+			InfrastructureRef: infraRef,
+		},
+	}
+
+	// Set the labels from machinePool.Spec.Template.Labels as labels for the new Machine.
+	// Note: We can't just set `machinePool.Spec.Template.Labels` directly and thus "share" the labels
+	// map between Machine and machinePool.Spec.Template.Labels. This would mean that adding the
+	// MachinePoolNameLabel later on the Machine would also add the labels to machinePool.Spec.Template.Labels
+	// and thus modify the labels of the MachinePool.
+	for k, v := range mp.Spec.Template.Labels {
+		machine.Labels[k] = v
+	}
+
+	// Enforce that the MachinePoolNameLabel and ClusterNameLabel are present on the Machine.
+	machine.Labels[clusterv1.MachinePoolNameLabel] = capilabels.MustFormatValue(mp.Name)
+	machine.Labels[clusterv1.ClusterNameLabel] = mp.Spec.ClusterName
```

Comment:
Q: I assume that label propagation will be implemented as a follow-up, this requires some thinking WRT to https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html and how to keep things in sync (cc @ykakarap who did all the work for MP)